yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala

```diff
@@ -245,4 +247,55 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
       System.clearProperty("SPARK_YARN_MODE")
     }
   }
+
+  test("Obtain tokens For HiveMetastore") {
+    val hadoopConf = new Configuration()
+    hadoopConf.set("hive.metastore.kerberos.principal", "bob")
+    // thrift picks up on port 0 and bails out, without trying to talk to endpoint
+    hadoopConf.set("hive.metastore.uris", "http://localhost:0")
+    val util = new YarnSparkHadoopUtil
+    val e = intercept[InvocationTargetException] {
+      util.obtainTokenForHiveMetastoreInner(hadoopConf, "alice")
+    }
+    assertNestedHiveException(e)
+    // expect exception trapping code to unwind this hive-side exception
+    assertNestedHiveException(intercept[InvocationTargetException] {
+      util.obtainTokenForHiveMetastore(hadoopConf)
+    })
+  }
+
+  def assertNestedHiveException(e: InvocationTargetException): Throwable = {
+    val inner = e.getCause
+    if (inner == null) {
+      fail("No inner cause", e)
+    }
+    if (!inner.isInstanceOf[HiveException]) {
+      fail(s"Not a hive exception", inner)
+    }
+    inner
+  }
+
+  test("handleTokenIntrospectionFailure") {
+    val util = new YarnSparkHadoopUtil
+    // downgraded exceptions
+    util.handleTokenIntrospectionFailure("hive", new ClassNotFoundException("cnfe"))
```

Comment:
Or yet another option is to have the method that handles exception take a closure, instead of the current approach of a method that matches on an exception parameter. e.g.:

    def tryToGetTokens(service: String)(fn: () => Option[Token]): Option[Token] = {
      try {
        fn()
      } catch {
         ...
      }
    }

    def obtainTokenForHiveMetastore... = {
      tryToGetTokens("Hive") { obtainTokenForHiveMetastoreInner(...) }
    }

I mostly dislike that exception handling feels like it's scattered around. You have a catch block in one place, then match on the exception somewhere else, it makes it hard to see what's really being done in one look.